cmd/collector/app/zipkin/protov2.go

```diff
@@ -0,0 +1,211 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package zipkin
+
+import (
+	"encoding/binary"
+	"fmt"
+	"net"
+
+	model "github.com/jaegertracing/jaeger/model"
+	zipkinProto "github.com/jaegertracing/jaeger/proto-gen/zipkin"
+	"github.com/jaegertracing/jaeger/thrift-gen/zipkincore"
+)
+
+// Converts Zipkin Protobuf spans to Thrift model
+func protoSpansV2ToThrift(listOfSpans *zipkinProto.ListOfSpans) ([]*zipkincore.Span, error) {
+	tSpans := make([]*zipkincore.Span, 0, len(listOfSpans.Spans))
+	for _, span := range listOfSpans.Spans {
+		tSpan, err := protoSpanV2ToThrift(span)
+		if err != nil {
+			return nil, err
+		}
+		tSpans = append(tSpans, tSpan)
+	}
+	return tSpans, nil
+}
+
+func protoSpanV2ToThrift(s *zipkinProto.Span) (*zipkincore.Span, error) {
+	if len(s.Id) != model.TraceIDShortBytesLen {
+		return nil, fmt.Errorf("invalid length for Span ID")
+	}
+	id := binary.BigEndian.Uint64(s.Id)
+	traceID, err := traceIDFromBytes(s.TraceId)
+	if err != nil {
+		return nil, err
+	}
+	ts, d := int64(s.Timestamp), int64(s.Duration)
+	tSpan := &zipkincore.Span{
+		ID:        int64(id),
+		TraceID:   int64(traceID.Low),
+		Name:      s.Name,
+		Debug:     s.Debug,
+		Timestamp: &ts,
+		Duration:  &d,
+	}
+	if traceID.High != 0 {
+		help := int64(traceID.High)
+		tSpan.TraceIDHigh = &help
+	}
+
+	if len(s.ParentId) > 0 {
+		if len(s.ParentId) != model.TraceIDShortBytesLen {
+			return nil, fmt.Errorf("invalid length for parentId")
+		}
+		parentID := binary.BigEndian.Uint64(s.ParentId)
+		signed := int64(parentID)
+		tSpan.ParentID = &signed
+	}
+
+	var localE *zipkincore.Endpoint
+	if s.LocalEndpoint != nil {
+		localE, err = protoEndpointV2ToThrift(s.LocalEndpoint)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	for _, a := range s.Annotations {
+		tA := protoAnnoV2ToThrift(a, localE)
+		tSpan.Annotations = append(tSpan.Annotations, tA)
+	}
+
+	tSpan.BinaryAnnotations = append(tSpan.BinaryAnnotations, tagsToThrift(s.Tags, localE)...)
+	tSpan.Annotations = append(tSpan.Annotations, protoKindToThrift(ts, d, s.Kind, localE)...)
+
+	if s.RemoteEndpoint != nil {
+		rAddrAnno, err := protoRemoteEndpToAddrAnno(s.RemoteEndpoint, s.Kind)
+		if err != nil {
+			return nil, err
+		}
+		tSpan.BinaryAnnotations = append(tSpan.BinaryAnnotations, rAddrAnno)
+	}
+
+	// add local component to represent service name
+	// to_domain looks for a service name in all [bin]annotations
+	if localE != nil && len(tSpan.BinaryAnnotations) == 0 && len(tSpan.Annotations) == 0 {
+		tSpan.BinaryAnnotations = append(tSpan.BinaryAnnotations, &zipkincore.BinaryAnnotation{
+			Key:            zipkincore.LOCAL_COMPONENT,
+			Host:           localE,
+			AnnotationType: zipkincore.AnnotationType_STRING,
+		})
+	}
+	return tSpan, nil
+}
+
+func traceIDFromBytes(tid []byte) (model.TraceID, error) {
```

Comment:
Perhaps incorporate this into `func (t *TraceID) Unmarshal(data []byte) error` instead. In which case the `*BytesLen` constants can remain private